### ISSUE TITLE
[BUG] Run "rspamadm dmarc_report" as cronjob without date parameter

### ISSUE THREAD

**Issue by PHPGangsta** [rspamd/rspamd#4062]
Hi,

I'm using rspamd 3.1-1~bionic.

I had a cronjob each day to send DMARC reports, like this:

17 9 * * * /usr/bin/rspamadm dmarc_report

@RealSebFox and I were confiused because sometimes we received DMARC-reports from each other, sometimes not. It looked like not all DMARC reports were sent, but we were not able to find out why.

After hours of investigation and checking Redis-contents (`SMEMBERS dmarc_idx;20220201`, `SMEMBERS dmarc_idx;20220202`) we found the solution inside the "--help":

```
$ rspamadm dmarc_report --help
Usage: rspamadm dmarc_report [-h] [-c <cfg>] [-v] [-n] [-b <number>]
       [<YYYYMMDD>] ...

Dmarc reports sending tool

Arguments:
   date                       Date to process (today by default)

Options:
   -h, --help                 Show this help message and exit.
         -c <cfg>,            Path to config file (default: /etc/rspamd/rspamd.conf)
   --config <cfg>
   -v, --verbose              Enable dmarc specific logging
   -n, --no-opt               Do not reset reporting data/send reports
             -b <number>,     Send reports in batches up to <batch-size> messages (default: 10)
   --batch-size <number>

```

`Date to process (today by default)`

If you just call `rspamadm dmarc_report` in your cronjob, it sends the reports `for the current day`, which means if you run it at 09:17, it only sends reports for mails between 00:00 and 09:17...

We both were expecting that by default it sends the reports of the day before (that makes the most sense in a cronjob, then you cover the whole day).

With the current behaviour you have 2 possibilities:

1. Run the Cronjob at 23:59 to cover the whole day... But it's ugly if all people in this world send emails during one minute, It should be spread over the whole day (everyone should choose a random time), so the recipients don't have problems with their "receive-rate-limit".
2. Run the cronjob with the date of yesterday as parameter, like this:
`17 2 * * * /usr/bin/rspamadm dmarc_report -v $(date --date yesterday "+\%Y\%m\%d") >> /tmp/dmarc_report.log 2>&1`

This strange behaviour of `rspamadm dmarc_report` explains the problem in Issue #4053 .

I guess most people who have a cronjob for sending rspamd DMARC reports don't know about this behaviour, and only send for a few hours per day (because they just run `rspamadm dmarc_report` without a date parameter)

If the default date ("today") will not be changed, then I would definitely recommend to add this important information to the documentation, and even provide an example of a correct cronjob with the date of yesterday (see above).

Currently in the documentation it says:

`From Rspamd 3.0 you should use "rspamadm dmarc_report" tool called manually (e.g. via cron or systemd timers) to send reports, this should be done either daily or hourly depending on traffic.`

Kind regards
Michael

**Comment by RealSebFox** [comment]
Hello,

According to RFC7489:

```
   Report generators SHOULD, wherever possible, adhere to
   hour boundaries for the reporting period they are using.  For
   example, starting a per-day report at 00:00; starting per-hour
   reports at 00:00, 01:00, 02:00; etc.  Report generators using a
   24-hour report period are strongly encouraged to begin that period at
   00:00 UTC, regardless of local timezone or time of report production,
   in order to facilitate correlation.
```

I believe yesterday's date should be assumed. There is no good reason to send out reports for the current day.

My fife cents..

Thank you!

**Comment by slavkoja** [comment]
I use this from when i switch to 3.x version to force yesterday date:
```
# send DMARC reports
21  3 * * *     root /usr/bin/rspamadm dmarc_report $(/usr/bin/date -d "1 day ago" +\%Y\%m\%d)
```
But the reports has set it's range from "03:21 date-1" -- "3:21 date"...

Ad RFC and day/hour boundary: people will have problem with some email providers due traffic spikes, e.g. google will reject (permanent) mails at spikes (counting others mailboxes too), see details in mailop ML archive some weeks ago.

Anyway, this advice is not good and i do not see anyone to follow it. Just consider, what if the report will be temporary rejected or network issue happens or power outage (etc)? Its delivery haven't be repeated latter, as it will not be at proper boundary? Fortunately, it is SHOULD and not MUST...

**Comment by RealSebFox** [comment]
> Anyway, this advice is not good and i do not see anyone to follow it. Just consider, what if the report will be temporary rejected or network issue happens or power outage (etc)? Its delivery haven't be repeated latter, as it will not be at proper boundary? Fortunately, it is SHOULD and not MUST...

The advice is about the reporting-period, not about the sending time. Your reporting-period can span from 00:00 to 00:00, also when you send your reports at 03:21 UTC. Those are two different things. 

The thing about the reporting boundaries in RFC7489 is a good advice. I prefer to keep it, even though I wish to send my reports at 07:00 CET. I personally am happy with the cron job like the one suggested by @PHPGangsta in option #2, but I don't believe it's a good solution overall and for everybody. Most people don't understand the difference between the reporting-period and the sending time, after all..

**Comment by slavkoja** [comment]
> The advice is about the reporting-period, not about the sending time. Your reporting-period can span from 00:00 to 00:00, also when you send your reports at 03:21 UTC. Those are two different things.

Oh, i misread it then, sorry, i agree with reporting ;-)

But anyway, with above command, the report (send to my self after fake sender):
```xml
<date_range>
  <begin>1643682063</begin>
  <end>1643768463</end>
</date_range>
```
Human readable range (UTC, day.month.year hour:min):
```
01.02.2022 02:21 – 02.02.2022 02:21
```

**Comment by RealSebFox** [comment]
> Human readable range (UTC, day.month.year hour:min):
> 
> ```
> 01.02.2022 02:21 – 02.02.2022 02:21
> ```

You are right. I see the same thing. My reports are also not showing a reporting period from 00:00 until 00:00, even though I have set my cron job to run every 24 hours with the date of yesterday. 

This is absolutely inconsistent. This way it only works properly with an hourly reporting period, but the daily reporting period is the standard and it's - as you correctly stated - not always recommended to send at the full hour.

**Comment by vstakhov** [comment]
Isn't it really fixed in the head similarly to #4053?

**Comment by slavkoja** [comment]
I do not know, i am on latest version.

**Comment by RealSebFox** [comment]
> Isn't it really fixed in the head similarly to #4053?

We've been looking at that code and we don't see how it fixes this particular issue.

**Comment by slavkoja** [comment]
> Isn't it really fixed in the head similarly to #4053?

By my little Lua knowlege, the ``start_collection`` is still based on current time and that is bad. It have at least floored to hour start and in case daily reports to day start.

The current logic is little magic, IMO solution can be to introduce new command line option to set that it is daily report, or to set the desired interval (eg. hourly, every 2, 3, 4, 6, 8, 12 hours, daily, or simple by hours numbers) and then floor time value to this boundary, thus if somebody send every 6 hours, floor the value to 0:00, 6:00: 12:00 or 18:00.

That solution (with new option) is not perfect, as it is prone to bad combination of command line interval with current cronjob frequency. Thus i can imagine this guess algorithm:

1. set ``start_collection`` to current hour start (eg. 3.21 => 3.00)
2. get last collection time and try to guess interval from it:

   1. if ``start_collection`` - ``last_collection`` > 24h, it is daily report = floor ``start_collection`` to 0:00
   2. otherwise floor it to value, which is "repeatable", eg. ``start_collection`` - ``last_collection`` < 2h, it is hourly report, thus leave current value, if ``start_collection`` - ``last_collection`` < 6h, it is 4h report, etc.
   3. if ``start_collection`` - ``last_collection`` < 1h, skip processing

3. process pending reports from ``last_collection`` to ``start_collection``
4. store ``start_collection`` as ``last_collection``

As i "test" it only in my head, i can miss something, but by this algorithm, the reports will be always for whole hours and its interval will be given only by cronjob settings. It is mostly selfhealthing, if there will be longer period as expected (some outage), reports will be send for longer period, but then period will decrease, not immediate but it will.

**Comment by vstakhov** [comment]
Hum, probably something like `rspamadm dmarc_report --cron`?

**Comment by PHPGangsta** [comment]
Probably something like:
`rspamadm dmarc_report --last-hour`
`rspamadm dmarc_report --last-day`

The first can be used in an "hourly" cronjob. It will generate reports for the last full hour, from XX:00 to YY:00. For example:
21 * * * * /usr/bin/rspamadm dmarc_report --last-hour

The second can be used in a "daily" cronjob. It will generate reports for the last day, from 00:00 to 23:59
26 3 * * * * /usr/bin/rspamadm dmarc_report --last-day

For most people the "daily" cronjob should be used, even Google and other big players only send daily DMARC reports.

**Comment by vstakhov** [comment]
Yes, it looks reasonable and self-descriptive. I was also thinking about `--cron=daily` and `--cron=hourly` but your suggestion is better I think.

**Comment by slavkoja** [comment]
OK, but what about those, who sends reports eg. every 4 hours to split amount to smaller chunks?

**Comment by PHPGangsta** [comment]
@slavkoja I guess we shouldn't make it too complex. Most people will send daily (and should send daily).. Even Google, Yahoo, Microsoft and so on send one DMARC report per domain per day. I haven't see anyone sending smaller time-ranges. And for those guys we have the "hourly" alternative.
Do you know anyone who is not sending daily or hourly, but for example "every 4 hours"? 

There are already lots of recipients running into "recipient-ratelimit-reached" temp-rejects (at gmail mailboxes for example). If you send more than one report per day, you stress this "recipient-ratelimit-reached" even more. That's why it's suggested to send daily I guess (and not multiple times per day).

**Comment by slavkoja** [comment]
You cannot compare big companies, where the reports can be stored (and handled) separated eg. by destination domain, with rspamd's implementation, which keeps all domains in one place and process them all at once on one instance...

When i have to choose, if i will overload google or self, then i choose google, which is sending me the same report (by its ID) 2-4 times daily.

**Comment by vstakhov** [comment]
Ok, after some looking into the code I found that your suggestions are impossible to implement with the current architecture. Unfortunately, I have not enough time to rework it, so I would say that it won't be done.

The problem is that Rspamd stores redis keys as following: `domain_<DDMMYYYY>`. So the only way to send report is to select some date and send reports for it. There are no other options are available, sorry. And changing that would require to change both Rspamd `dmarc` plugin code and significantly change the `dmarc_report` logic. It will also break the existing reports so please bear in mind that I would be reluctant to changes here due to POLA.

**Comment by vstakhov** [comment]
Or I can probably just fix `--yesterday` behaviour instead...

**Comment by slavkoja** [comment]
IMO, fixing yesterday can be enough, in other case one can simple setup cronjob on hour boundary

**Comment by RealSebFox** [comment]
> Or I can probably just fix `--yesterday` behaviour instead...

If the reporting period for `--yesterday` would only span from yesterday 00:00 UTC until today 00:00 UTC, regardless of the sending time of the cron job, that would already be a great improvement. 

Not to say rspamd is not great. I believe it generally is a great tool!

**Comment by RealSebFox** [comment]
PS: Please don't forget to explain in the user manual, that `rspamadm dmarc_report` without date, executed on a daily base, is insufficient to send out all reports. It's an important point for users to understand.

> I guess most people who have a cronjob for sending rspamd DMARC reports don't know about this behaviour, and only send for a few hours per day (because they just run `rspamadm dmarc_report` without a date parameter)

**Comment by vstakhov** [comment]
Can anyone test the current behaviour please?

**Comment by RealSebFox** [comment]
> Can anyone test the current behaviour please?

I'm more than happy to do so with the current (latest) version we have online, but please be more specific in detailing the test. I've explained exactly what we did and what we saw in #4053: I've been observing redis and verbose log files for several hours, from 00:00 until past the cron job time. That was quite a long night! :-D

If you can tell me exactly what is missing from #4053, I can test it this week.

**Comment by PHPGangsta** [comment]
I don't compile rspamd myself, I'm using the Ubuntu/Debian packages from rspamd.com (currently 3.1-1~focal amd64).

To test the new behaviour I have to compile it myself and run it for a few days, right? I'm not sure if I have time for that 😢 

Or can I apply the changes from this commit manually to the .lua file and just restart rspamd?
https://github.com/rspamd/rspamd/commit/47ff44ea1509ea7882377bf14c541b799eb4cf62

**Comment by RealSebFox** [comment]
Just to recap: Because this feature requires incoming mail streams to work, it can - as far as i can see - only be properly tested on an active, live, running system. Most of us - mail administrators - don't compile the software we use ourselves, but we rather take it from existing repositories, especially on the live systems with actual mail streams. 

Doesn't it generally make sense to have a test-repository ("beta" or "nightly"), so the community is better able to provide feedback?

**Comment by PHPGangsta** [comment]
I updated to 3.2 two days ago and removed the "date" parameter from the cronjob, now it's just:

`17 2 * * * /usr/bin/rspamadm dmarc_report -v >> /tmp/dmarc_report.log 2>&1`

But: The date-range is a bit wrong:

```
    <date_range>
      <begin>1648940400</begin>
      <end>1649026800</end>
    </date_range>
```

According to the RFC it should be from 00:00 UTC to 00:00 UTC. But now it's from 23:00 UTC to 23:00 UTC.

My server is running in CEST timezone at the momemt, which is GMT+2... Not sure if it has to do with my selected timezone in the Operating System.

**Comment by PHPGangsta** [comment]
@slavkoja Thanks for your confirmation of the problems. I guess @vstakhov will find a solution for calculating the start-timestamp and end-timestamp in UTC.

You are right that the data in redis seems to be stored with YYYYMMDD, but I hope that the used date is UTC when calculating YYYYMMDD.

You are also right, when the reports have been sent today morning rspamd wrote to the logfile:
`Process date 20220403`
and the last 2 lines were:
```
set last report date to 1649113200
Reporting collection has finished 1 dates processed, 13 reports: 13 completed, 0 failed
```
With the timestamp being 04.04.2022 23:00 UTC.

I did not check if rspamd fetched the data of 20220403 or 20220404 when generating the reports, you might be right that it has used the redis-data of 20220403.
